### PR TITLE
Introduce (Get|Select|Exec)Builder

### PR DESCRIPTION
**Pull request by lieut-data**
#### Summary
When using squirrel, we're often forced to write unnecessary boilerplate to first build the query and then execute it:

```go
sql, args, err := s.getQueryBuilder().
    Select("*").
    From("Threads").
    Where(sq.Eq{"PostId": id}).
    ToSql()
if err != nil {
    return nil, errors.Wrapf(err, "failed to build query")
}

var thread model.Thread
err := s.GetMasterX().Get(&thread, sql, args...)
if err != nil {
    return nil, errors.Wrapf(err, "failed to get thread")
}
```

but we can inline the query building and error handling, streamlining this simply be:
```go
query := s.getQueryBuilder().
    Select("*").
    From("Threads").
    Where(sq.Eq{"PostId": id})

var thread model.Thread
err := s.GetMasterX().GetBuilder(&thread, query)
if err != nil {
    return nil, errors.Wrapf(err, "failed to get thread")
}
```

Prototype this in the `ThreadStore` by extending the `sqlxDBWrapper`.  Note that squirrel itself supports something like this with a `query.RunWith(s.GetMasterX())`, but adopting this would change the semantics of our struct parsing (from sqlx to squirrel), so this change seeks to limit the potential of unexpected side effects.

#### Ticket Link
None.

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: store/sqlstore/sqlx_wrapper.go [hunk 1]
```diff
@@ -35,17 +35,24 @@ func (w *StoreTestWrapper) DriverName() string {
...
 
+type Builder interface {
+	ToSql() (string, []interface{}, error)
+}
+
 // sqlxExecutor exposes sqlx operations. It is used to enable some internal store methods to
 // accept both transactions (*sqlxTxWrapper) and common db handlers (*sqlxDbWrapper).
 type sqlxExecutor interface {
 	Get(dest interface{}, query string, args ...interface{}) error
+	GetBuilder(dest interface{}, builder Builder) error
```
Comment (lieut-data) [inline store/sqlstore/sqlx_wrapper.go:46]: I've only extended a handful of methods -- the ones I needed in the `ThreadStore` -- but it's clearly easy to do if we need them elsewhere.

File: store/sqlstore/thread_store.go [hunk 2]
```diff
@@ -497,20 +461,15 @@ func (s *SqlThreadStore) GetThreadForUser(teamId string, threadMembership *model
 	}
 
 	var thread JoinedThread
-	query, threadArgs, err := s.getQueryBuilder().
+	query := s.getQueryBuilder().
 		Select("Threads.*, Posts.*").
 		From("Threads").
-		Column(sq.Alias(sq.Expr(unreadRepliesQuery), "UnreadReplies")).
+		Column(sq.Alias(sq.Expr(unreadRepliesQuery, unreadRepliesArgs...), "UnreadReplies")).
```
Comment (lieut-data) [inline store/sqlstore/thread_store.go:467]: This is a nice improvement as well.

File: store/sqlstore/thread_store.go [hunk 3]
```diff
@@ -987,23 +906,21 @@ func (s *SqlThreadStore) DeleteOrphanedRows(limit int) (deleted int64, err error
 }
 
 // return number of unread replies for a single thread
-func (s *SqlThreadStore) GetThreadUnreadReplyCount(threadMembership *model.ThreadMembership) (unreadReplies int64, err error) {
-	query, args, err := s.getQueryBuilder().
+func (s *SqlThreadStore) GetThreadUnreadReplyCount(threadMembership *model.ThreadMembership) (int64, error) {
```
Comment (lieut-data) [inline store/sqlstore/thread_store.go:909]: Removed the name return params in a separate commit to avoid the clash with `err`.

**Review by agnivade — APPROVED** [review]
Nice work!

Might I suggest to hold off merging this as this clashes with @ashishbhate's work in https://github.com/mattermost/mattermost-server/pull/20034?

File: store/sqlstore/thread_store.go (hunk 2 above)
Comment (lieut-data) [reply store/sqlstore/thread_store.go:485]: So `sq.Expr` creates a `Sqlizer` embedding its own `sql` and `args`. `sq.Alias` then holds onto the `Sqlizer`, and `Column` creates a nested `Sqlizer`. When it comes time to finally build the query, `Sqlizer`'s `ToSQL` extracts the args from the embedded `Sqlizer`:

https://github.com/lieut-data/squirrel/blob/6ba7e8d947599922316c8b509ba57512659959c6/part.go#L17-L30

The `args` array gets built up as we process all of these embedded `Sqlizer` statements:

https://github.com/lieut-data/squirrel/blob/6ba7e8d947599922316c8b509ba57512659959c6/select.go#L88-L93

Long story short, this works:
```go
package main

import (
	"fmt"
	"testing"

	sq "github.com/Masterminds/squirrel"
)

func TestExpr(t *testing.T) {
}

func ExampleExpr() {
	subQuery, subqueryArgs, _ := sq.StatementBuilder.PlaceholderFormat(sq.Question).
		Select("COUNT(Posts.Id)").
		From("Posts").
		Where(sq.NotEq{"DeleteAt": 0}).
		ToSql()

	query, args, _ := sq.StatementBuilder.PlaceholderFormat(sq.Question).
		Select("PostId").
		Column(sq.Alias(sq.Expr(subQuery, subqueryArgs...), "NumPosts")).
		From("Threads").
		Where(sq.Eq{"PostId": "post_id"}).
		ToSql()

	fmt.Println(query)
	fmt.Println(args)

	// Output: SELECT PostId, (SELECT COUNT(Posts.Id) FROM Posts WHERE DeleteAt <> ?) AS NumPosts FROM Threads WHERE PostId = ?
	// [0 post_id]
}
```

```
> go test -v
=== RUN   TestExpr
--- PASS: TestExpr (0.00s)
=== RUN   ExampleExpr
--- PASS: ExampleExpr (0.00s)
PASS
ok      test    0.093s
```

The need to "render" the subquery feels like something else we could add to our fork of squirrel -- it should suffice to just write something like the following (or even inlining the subquery directly!):
```go
subQuery := sq.StatementBuilder.PlaceholderFormat(sq.Question).
        Select("COUNT(Posts.Id)").
        From("Posts").
        Where(sq.NotEq{"DeleteAt": 0})

query, args, _ := sq.StatementBuilder.PlaceholderFormat(sq.Question).
        Select("PostId").
        Column(sq.Alias(sq.Expr(subQuery), "NumPosts")).
        From("Threads").
        Where(sq.Eq{"PostId": "post_id"}).
        ToSql()

```

Comment (lieut-data) [reply store/sqlstore/thread_store.go:485]: Another way of saying this, is that it aggregates args just like `.Where(...)` adds args to the final array.

**Comment by lieut-data** [comment]
> Might I suggest to hold off merging this as this clashes with @ashishbhate's work in https://github.com/mattermost/mattermost-server/pull/20034?

Yes, absolutely! No rush whatsoever on this, and happy to own resolving conflicts.

File: store/sqlstore/thread_store.go (hunk 2 above)
Comment (agnivade) [reply store/sqlstore/thread_store.go:485]: >  (or even inlining the subquery directly!)

ooh nice I like this!

**Comment by lieut-data** [comment]
@jgilliam17, no hurry on this one, but could I get your help running and reviewing e2e tests? There should be zero functional changes from this PR.

**Comment by agnivade** [comment]
Actually, since @ashishbhate's work will be pushed back, let's get this stuff merged. :+1:

**Comment by jgilliam17** [comment]
test running [here](https://automation-dashboard.vercel.app/cycles/452841d7-29b3-404d-835c-3fa384e83857)

**Comment by jgilliam17** [comment]
cycle failed, updating and re-running the test

**Comment by jgilliam17** [comment]
Thanks @lieut-data 
Here is the [E2E report,](https://mattermost-cypress-report.s3.amazonaws.com/190065-381af56-onprem-ent-server-pr-20029-mattermost/mm-ee-test381af56/mochawesome.html) has 8 failures, vs. [master](https://mattermost-cypress-report.s3.amazonaws.com/190084-onprem-ent-master-mm-ee-test:test/mochawesome.html) 2 
MM-T1859 Bot is kept active when owner is disabled
MM-T2181 Custom emoji - add large
MM-T2185 Custom emoji - renders immediately for other user Custom emoji - renders after logging out and back in
Flaky not related to PR - MM-T1270 UP - Edit message with attachment but no text
MM-T249 Mark GM post as unread
MM-T2368 Fixed width
MM-T952 Reactivating a user results in them showing up in the normal spot in the list, without the `Deactivated` label.
Flaky not related to PR  - MM-T1786 Dismissing the toast using Jump to

I'll re-test the failures manually to confirm, as soon as the cluster is available.

**Comment by lieut-data** [comment]
@jgilliam17, any luck identifying the source of those test failures?

**Comment by jgilliam17** [comment]
Apologies, this slipped off my radar due to release testing. Let me update and re-run.

**Comment by jgilliam17** [comment]
Not sure if we are having issues with Cypress right now, but I am getting more failures on re-runs.
[E2E report ](https://mattermost-cypress-report.s3.amazonaws.com/194871-1e85541-onprem-ent-server-pr-20029-mm-ee-test:1e85541/mochawesome.html)
Looking into

**Comment by jgilliam17** [comment]
@lieut-data this looks good to merge
Latest [e2e report ](https://mattermost-cypress-report.s3.amazonaws.com/194966-babc433-onprem-ent-server-pr-20029-mm-ee-test:babc433/mochawesome.html)  has only 2 failures that differ from [master](https://mattermost-cypress-report.s3.amazonaws.com/194670-onprem-ent-master-mm-ee-test:test/mochawesome.html) - no concern from QA standpoint, the same 2 tests are intermittently failing on other PRs.

**Comment by lieut-data** [comment]
Hey @amyblais! Any specific reason to cherry-pick this to `cloud`? It's mostly just a code cleanup item that can ship organically.

**Comment by amyblais** [comment]
Sounds good!

**Pull request merged**